### DISCUSSION TITLE
cv2 diaply to wslg

### DISCUSSION THREAD

**Discussion by sharpened-flat** [microsoft/wslg#247] [Q&A]
Hi team, excellent feature thank you! One question though, cv2 module (for example in python3) - can I make it show to wslg?

import cv2
image = cv2.imread("image.jpg")
cv2.imshow("test",image)

can we? Till now I had to install a seprate Xserver and declare by "export DISPLAY=192.168.50.1:0.0" something like that...

Kevin

**Comment by onomatopellan** [comment]
I just tried and it just works. Just make sure the env variable remains `DISPLAY=:0` .

**Reply by sharpened-flat** [reply]
are you sure? I still have errors (I am not running Xserver, hoping cv2 can directly display to wslg)

fryhard@SLFAT:/tmp$ export DISPLAY=:0
fryhard@SLFAT:/tmp$ python3
Python 3.8.5 (default, Jan 27 2021, 15:41:15)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import cv2
>>> image = cv2.imread("body_front.jpg")
>>> cv2.imshow("test",image)
qt.qpa.xcb: could not connect to display :0
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "/home/fryhard/.local/lib/python3.8/site-packages/cv2/qt/plugins" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: xcb.

Aborted

**Reply by onomatopellan** [reply]
From inside the interpreter you probably lack some library path. I just did run `pip install opencv-python` and it worked from the command line. The distro I'm running is Ubuntu on Windows Community Preview.

![2021-05-30](https://user-images.githubusercontent.com/599142/120103909-31af8e80-c152-11eb-946e-343f3d063cbe.png)

**Comment by sharpened-flat** [comment]
Thanks for your quick response but I still get the same error.
This time I tested in the same envrionment as yours. I downloaded Ubuntu on Windows Community Preview (ver 21.04). These are the commands I executed till I got the error:

sudo apt update
sudo apt upgrade
sudo apt install python3-pip
pip3 install opencv-python
export DISPLAY=:0
python3 test.py

and the result:

fryhard@SLFAT:/mnt/c/Users/fryha/dropbox/sharpened_flat/labs/background-remover$ python3 test.py
qt.qpa.xcb: could not connect to display :0
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "/home/fryhard/.local/lib/python3.9/site-packages/cv2/qt/plugins" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: xcb.

BTW my laptop is Microsoft Surfbook 3

**Reply by onomatopellan** [reply]
Try installing and running a Qt application like vlc. Does it work with WSLg?

If it still doesn't work try to get more info from the error with:
```
export QT_DEBUG_PLUGINS=1
python3 test.py
```

**Reply by onomatopellan** [reply]
I just fixed a similar problem. I was trying to run cv2 inside the Systemdistro which contains Cbl-Mariner as distro and lacks the packages to display anything with cv2.

So I did run my test file with debug enabled
`QT_DEBUG_PLUGINS=1 python3 test.py`

and this is the output:

```
"Cannot load library /usr/lib/python3.7/site-packages/cv2/qt/plugins/platforms/libqxcb.so: (libSM.so.6: cannot open shared object file: No such file or directory)"
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "/usr/lib/python3.7/site-packages/cv2/qt/plugins" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```

So I installed the libSM package with `tdnf install libSM` and finally it worked.

**Reply by sharpened-flat** [reply]
my case is differnt. I have no issue of missing libSM library.

I tried all of the suggestions you made so far.
1. install vlc - no luck
2. install libSM - already installed
3. analysing the debug output - only the meaning part is same as before - qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "/home/fryhard/.local/lib/python3.9/site-packages/cv2/qt/plugins"

Next step - if you are keen to help me find the root cause, let me suggest a Zoom session in which you remotly control my wsl2 for torubleshoot. I am in Tokyo at GMT+9.

Thanks,
Kevin

**Reply by onomatopellan** [reply]
Sorry, I've run out of ideas. At this point we can consider it an issue with WSLg so you should post it in https://github.com/microsoft/wslg/issues providing all the info you can, [detailed logs](https://github.com/Microsoft/WSL/blob/master/CONTRIBUTING.md#8-detailed-logs) included.

**Comment by onomatopellan** [comment]
I had to ask this sooner, but which Windows build are your running? (`winver.exe`)

**Comment by sharpened-flat** [comment]
Windows 10 Dev 21390.1010

 
发件人: onomatopellan ***@***.***> 

发送时间: 2021年6月15日 3:09
收件人: microsoft/wslg ***@***.***>
抄送: Info ***@***.***>; Author ***@***.***>
主题: Re: [microsoft/wslg] cv2 diaply to wslg (#247)

 
I had to ask this sooner, but which Windows build are your running? (winver.exe)

**Reply by onomatopellan** [reply]
The only significant difference with my setup is my screen is 1920x1080 with 100% scaling while yours has probably a higher DPI. You can try upgrading your WSLg to the latest version 1.0.24 which have some Hi-DPI fixes. You can download it from here: https://github.com/microsoft/wslg/releases/download/v1.0.24/wsl_graphics_support_x64.msi

**Reply by sharpened-flat** [reply]
It works finally. I can see a new Linux window pop up when cv2.imshow() is invoked. Still a minor issue on cv2.waitKey(0) where key pressed but window not closed, but I am happy that I dont need to rely on Xserver any more!

Thanks for your support.
Kevin
